File: app/src/main/java/com/cottondroid/todoachieved/task/TaskRepository.java
package com.cottondroid.todoachieved.task;

import com.cottondroid.todoachieved.task.model.TodoTask;
import com.cottondroid.todoachieved.task.model.TodoTaskDao;
import com.cottondroid.todoachieved.task.network.TaskUpdateListener;
import com.cottondroid.todoachieved.task.network.TodoTaskService;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;

import static com.cottondroid.todoachieved.di.ApplicationModule.IO_SCHEDULER;
import static com.cottondroid.todoachieved.di.ApplicationModule.UI_SCHEDULER;

@Singleton
public class TaskRepository {

    private final TodoTaskDao taskDao;
    private final TodoTaskService todoTaskService;
    private final Scheduler uiScheduler;
    private final Scheduler ioScheduler;

    @Inject
    public TaskRepository(TodoTaskDao taskDao,
                          TodoTaskService todoTaskService,
                          @Named(UI_SCHEDULER) Scheduler uiScheduler,
                          @Named(IO_SCHEDULER) Scheduler ioScheduler) {
        this.taskDao = taskDao;
        this.todoTaskService = todoTaskService;
        this.uiScheduler = uiScheduler;
        this.ioScheduler = ioScheduler;
    }

    public Flowable<List<TodoTask>> getTasks() {
        return taskDao.list()
                .subscribeOn(ioScheduler)
                .observeOn(uiScheduler);
    }

    public void registerToTaskUpdates(final TaskUpdateListener taskListener) {
        taskDao.getLastServerId()
                .defaultIfEmpty("")
                .subscribeOn(ioScheduler)
                .observeOn(uiScheduler).subscribe(new Consumer<String>() {
            @Override
            public void accept(String lastServerId) throws Exception {
                todoTaskService.registerToTaskUpdates(taskListener, lastServerId);
            }
        });
    }

    public void unregisterFromTaskUpdates(TaskUpdateListener taskListener) {
        todoTaskService.unregisterFromTaskUpdates(taskListener);
    }

    public Single<TodoTask> saveTask(final TodoTask todoTask) {
        return taskDao.insertOrReplace(todoTask)
                .flatMap(new Function<Long, SingleSource<TodoTask>>() {
                    @Override
                    public SingleSource<TodoTask> apply(Long taskId) throws Exception {
                        return taskDao.load(taskId);
                    }
                })
                .doOnSuccess(new Consumer<TodoTask>() {
                    @Override
                    public void accept(TodoTask task) throws Exception {
                        todoTaskService.populateServerValues(task);
                        taskDao.updateSync(task);
                        todoTaskService.saveTask(task);
                    }
                })
                .subscribeOn(ioScheduler)
                .observeOn(uiScheduler);
    }

    public Single<TodoTask> loadTask(final long taskId) {
        return taskDao.load(taskId)
                .subscribeOn(ioScheduler)
                .observeOn(uiScheduler);
    }
}

File: app/src/main/java/com/cottondroid/todoachieved/task/TaskRepository.kt
package com.cottondroid.todoachieved.task

import com.cottondroid.todoachieved.task.model.TodoTask
import com.cottondroid.todoachieved.task.model.TodoTaskDao
import com.cottondroid.todoachieved.task.network.TaskUpdateListener
import com.cottondroid.todoachieved.task.network.TodoTaskService
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TaskRepository @Inject constructor(
        private val taskDao: TodoTaskDao,
        private val todoTaskService: TodoTaskService
) {
    private val disposable = CompositeDisposable()
    val tasks: Flowable<List<TodoTask>>
        get() = taskDao.list()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())

    fun registerToTaskUpdates(taskListener: TaskUpdateListener) {
        disposable.add(
                taskDao.lastServerId
                        .defaultIfEmpty("")
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.io())
                        .subscribe { lastServerId ->
                            todoTaskService.registerToTaskUpdates(taskListener, lastServerId)
                        }
        )
    }

    fun unregisterFromTaskUpdates(taskListener: TaskUpdateListener) {
        todoTaskService.unregisterFromTaskUpdates(taskListener)
        disposable.clear()
    }

    fun saveTask(todoTask: TodoTask): Single<TodoTask> {
        return taskDao.insertOrReplace(todoTask)
                .flatMap { taskId ->
                    taskDao.load(taskId)
                }
                .doOnSuccess { task ->
                    todoTaskService.populateServerValues(task)
                    taskDao.updateSync(task)
                    todoTaskService.saveTask(task)
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

    fun loadTask(taskId: Long): Single<TodoTask> {
        return taskDao.load(taskId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
    }

}